Refactor the decl scope loop to consolidate modifier handling.

jonmeow opened the pull request:
Building on #3462, trying to make the flow easier to see, also making a little more use of macros for boilerplate modifier handling.

jonmeow committed:
Refactor the decl scope loop to consolidate modifier handling.

jonmeow committed:
Refactor the decl scope loop to consolidate modifier handling.

josh11b committed:
Abbreviate "representation" -> "repr" (#3464)

Specifically using these abbreviations:
* InitRepr: "initializing representation"
* ObjectRepr: "object representation"
* ValueRepr: "value representation"

As discussed in [#toolchain
discord](https://discord.com/channels/655572317891461132/655578254970716160/1182086098470572143)
and now documented in the [list of
abbreviations](https://docs.google.com/document/d/1RRYMm42osyqhI2LyjrjockYCutQ5dOf8Abu50kTrkX0/edit?resourcekey=0-kHyqOESbOHmzZphUbtLrTw#heading=h.pph7i5m5un7q).

zygoloid commented on toolchain/parse/handle_decl_scope_loop.cpp:
```diff
@@ -2,12 +2,54 @@
+      context.PushState(State::Package);
+      return true;
+    }
+    default: {
```
Optional: to reduce indentation, consider `break;`ing here and putting the `default` case after the loop.

jonmeow replied:
My leaning was this direction to keep everything in the switch, partly because `default` can't simply be removed. [I think it's visually more consistent]

zygoloid commented on toolchain/parse/handle_decl_scope_loop.cpp:
```diff
@@ -25,81 +67,192 @@ static auto OutputInvalidParseSubtree(Context& context, int32_t subtree_start)
-// Handles an unrecognized declaration.
-static auto HandleUnrecognizedDecl(Context& context, int32_t subtree_start)
-    -> void {
+// Reports an error and outputs an invalid parse tree.
```
I had trouble understanding what "outputs an invalid parse tree" means.

It seems like the purpose of `OutputInvalidParseSubtree` is to recover from an error by skipping to a good stopping place and fixing up the parse tree to match. I don't get that from the name. Maybe something more like `FinishAndSkipInvalidDecl` would be clearer?

jonmeow replied:
Done

zygoloid commented on toolchain/parse/handle_decl_scope_loop.cpp:
```diff
@@ -35,71 +77,182 @@ static auto HandleUnrecognizedDecl(Context& context, int32_t subtree_start)
-    case Lex::TokenKind::Import: {
-      context.PushState(State::Import);
-      return;
+
+    case Lex::TokenKind::Extend: {
+      if (TokenIsModifierOrIntroducer(
+              context.PositionKind(Lookahead::NextToken))) {
+        // `extend` is considered a declaration modifier if it is followed by
+        // another modifier or an introducer.
+        return false;
+      } else {
+        // TODO: Treat this `extend` token as a declaration introducer
+        HandleUnrecognizedDecl(context, state.subtree_start);
+        return true;
+      }
     }
-    case Lex::TokenKind::Library: {
-      context.PushState(State::Library);
-      return;
+
+    case Lex::TokenKind::Impl: {
+      if (TokenIsModifierOrIntroducer(
+              context.PositionKind(Lookahead::NextToken))) {
+        // `impl` is considered a declaration modifier if it is followed by
+        // another modifier or an introducer.
+        return false;
+      } else {
+        // TODO: Treat this `impl` token as a declaration introducer
+        HandleUnrecognizedDecl(context, state.subtree_start);
+        return true;
+      }
     }
```
It seems like there's still duplication here; can we check for the ambiguous cases first somehow? In all three cases, if a token is both an introducer and a modifier, the logic is the same: it's a modifier if the next token is a modifier or an introducer.

jonmeow replied:
Refactored a little for this.

jonmeow committed:
Addressing comments

jonmeow committed:
Addressing comments

jonmeow committed:
Switch modifier/declaration order

jonmeow committed:
Switch modifier/declaration order

zygoloid committed:
Move `llvm::Initialize*` calls to `main`. (#3449)

Per their documentation, the `llvm::Initialize*` functions are only
supposed to be called by the main program, not by a library like
toolchain/codegen. Fixes a hang due to a data race in multithreaded
autoupdate.

Add a utility class `Carbon::InitLLVM` to do the common LLVM
initialization shared by all Carbon tools, optionally including
initializing the LLVM targets. Because the LLVM targets add a lot of
binary size, only initialize them for binaries that opt in by depending
on a new target `//common:all_llvm_targets`.

Also fix `//explorer:file_test` and `//explorer:file_test.trace` to
share a binary rather than linking an identical binary twice.

---------

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Rename `Name` -> `IdentifierName` given that we have several other kinds of parse nodes that represent names. (#3453)

Factor out common checking handling for the different kinds of
unqualified names

zygoloid committed:
Factor out common pattern of trying to complete a type then falling back to an error type if that fails. (#3454)

As requested in [review of
#3450](https://github.com/carbon-language/carbon-lang/pull/3450#discussion_r1414220143).

josh11b committed:
Drop `GetDeclContext` from parse, errors will be diagnosed in check (#3468)

The criteria being enforced is not correct since `final` and `default`
functions in interfaces have definitions.

zygoloid commented on toolchain/parse/handle_decl_scope_loop.cpp:
```diff
@@ -25,81 +68,185 @@ static auto OutputInvalidParseSubtree(Context& context, int32_t subtree_start)
+// Handles `base` as a declaration.
+static auto HandleBaseAsDecl(Context& context, Context::StateStackEntry state)
+    -> void {
+  // `base` may be followed by:
+  // - a colon
+  //   => assume it is an introducer, as in `extend base: BaseType;`.
+  // - a modifier or an introducer
+  //   => assume it is a modifier, as in `base class`; which is handled
+  //      by falling through to the next case.
+  // Anything else is an error.
```
I think this comment is out of date (at least the modifier part and "falling through")

jonmeow replied:
Cleaning up the comment.

zygoloid commented on toolchain/parse/handle_decl_scope_loop.cpp:
```diff
@@ -110,164 +257,17 @@ auto HandleDeclScopeLoop(Context& context) -> void {
-  };
-
   bool saw_modifier = false;
   while (true) {
```
```suggestion
```

zygoloid commented on toolchain/parse/handle_decl_scope_loop.cpp:
```diff
@@ -110,164 +257,17 @@ auto HandleDeclScopeLoop(Context& context) -> void {
+    } else {
+      HandleUnrecognizedDecl(context, state.subtree_start);
+      return;
     }
   }
```
```suggestion
```

zygoloid commented on toolchain/parse/handle_decl_scope_loop.cpp:
```diff
@@ -110,164 +257,17 @@ auto HandleDeclScopeLoop(Context& context) -> void {
+      return;
+    } else {
+      HandleUnrecognizedDecl(context, state.subtree_start);
+      return;
```
I  think this would be a bit simpler if you moved the handling of the declaration out of the loop, eg:
```suggestion
  while (TryHandleAsModifier(context, context.PositionKind())) {
    saw_modifier = true;
  }
  if (!TryHandleAsDecl(context, state, saw_modifier, context.PositionKind())) {
    HandleUnrecognizedDecl(context, state.subtree_start);
  }
```
(Suggestion split in three parts to work around github limitations.)

zygoloid reviewed (approved):
Thanks!

jonmeow committed:
Apply suggestions from code review

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

jonmeow committed:
Addressing comments

jonmeow commented:
Thanks, I'd totally overlooked the ability to shift the `while`.
